src/protocolsupport/protocol/packet/middle/clientbound/play/MiddleDeclareRecipes.java

```diff
@@ -1,22 +1,226 @@
 package protocolsupport.protocol.packet.middle.clientbound.play;
 
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.Map;
+
 import io.netty.buffer.ByteBuf;
+import protocolsupport.api.ProtocolVersion;
 import protocolsupport.protocol.ConnectionImpl;
 import protocolsupport.protocol.packet.middle.ClientBoundMiddlePacket;
+import protocolsupport.protocol.serializer.ItemStackSerializer;
 import protocolsupport.protocol.serializer.MiscSerializer;
+import protocolsupport.protocol.serializer.StringSerializer;
+import protocolsupport.protocol.serializer.VarNumberSerializer;
+import protocolsupport.protocol.utils.ProtocolVersionsHelper;
+import protocolsupport.protocol.utils.i18n.I18NData;
+import protocolsupport.protocol.utils.types.NetworkItemStack;
 
 public abstract class MiddleDeclareRecipes extends ClientBoundMiddlePacket {
 
 	public MiddleDeclareRecipes(ConnectionImpl connection) {
 		super(connection);
 	}
 
-	//TODO: structure
-	protected ByteBuf data;
+	protected static final Map<ProtocolVersion, ByteBuf> cachedBuffers = Collections.synchronizedMap(new EnumMap<>(ProtocolVersion.class));
+
+	protected Recipe[] recipes;
 
 	@Override
 	public void readFromServerData(ByteBuf serverdata) {
-		data = MiscSerializer.readAllBytesSlice(serverdata);
+		if (cachedBuffers.containsKey(connection.getVersion())) {
+			MiscSerializer.readAllBytesSlice(serverdata);
+		} else {
+			int count = VarNumberSerializer.readVarInt(serverdata);
+			recipes = new Recipe[count];
+			for (int i = 0; i < count; i++) {
+				String id = StringSerializer.readString(serverdata, ProtocolVersionsHelper.LATEST_PC);
+				RecipeType type = RecipeType.valueOf(StringSerializer.readString(serverdata, ProtocolVersionsHelper.LATEST_PC));
```

Comment:
readString.toUpperCase() and have the enum in capital letters. (This is a style preference)